developerportal/apps/events/models.py

```diff
@@ -132,26 +148,120 @@ def can_create_at(cls, parent):
+                overall_date_q = date_q
+            else:
+                overall_date_q.add(date_q, Q.OR)
+
+        # Finally, ensure we don't include past events here (ie, same month as
+        # selected but before today)
+        overall_date_q.add(default_future_events_q, Q.AND)
+        return overall_date_q
+
+    def get_upcoming_events(self, request):
+        """Return filtered future events in chronological order"""
+        # These are not paginated but ARE filtered
+
+        countries = request.GET.getlist(COUNTRY_QUERYSTRING_KEY)
+        years_months = request.GET.getlist(YEAR_MONTH_QUERYSTRING_KEY)
+        topics = request.GET.getlist(TOPIC_QUERYSTRING_KEY)
+
+        countries_q = build_complex_filtering_query_from_query_params(
+            query_syntax="country", params=countries
+        )
+        topics_q = build_complex_filtering_query_from_query_params(
+            query_syntax="topics__topic__slug", params=topics
+        )
+
+        # year_months need splitting to make them work
+        date_q = self._build_date_q(years_months)
+
+        combined_q = Q()
+        if countries_q:
+            combined_q.add(countries_q, Q.AND)
+        if date_q:
+            combined_q.add(date_q, Q.AND)
+        if topics_q:
+            combined_q.add(topics_q, Q.AND)
+
+        # Combined_q will always have something because it includes
+        # the start_date__gte test
+        events = get_combined_events(self, q_object=combined_q)
+
+        return events
+
+    def get_past_events(self, request):
+        """Return paginated past events in reverse chronological order,
+        plus a count of how many there are in total
+        """
+        past_events = get_combined_events(
+            self, reverse=True, start_date__lt=tz_now().date()
         )
+        total_past_events = len(past_events)
+
+        past_events = paginate_resources(
+            past_events,
+            page_ref=request.GET.get(PAGINATION_QUERYSTRING_KEY),
+            per_page=self.PAST_EVENTS_PER_PAGE,
+        )
+        return past_events, total_past_events
+
+    def get_relevant_countries(self):
+        # Relevant here means a country that a published Event is or was in
+        raw_countries = set(
+            event.country
+            for event in Event.published_objects.filter(start_date__gte=tz_now().date())
+            if event.country
+        )
+
+        return sorted(
+            [{"code": country.code, "name": country.name} for country in raw_countries],
+            key=lambda x: x["code"],
+        )
+
+    def get_relevant_dates(self):
+        # Relevant here means a date for a published *future* event
+        # TODO: would be good to cache this for short period of time
+        raw_events = get_combined_events(self, start_date__gte=tz_now().date())
+        return sorted([event.start_date for event in raw_events])
 
     def get_filters(self):
         from ..topics.models import Topic
```

Comment:
Not your line, but do you know why the local import Circular dependencies? I suppose we should drop a comment about that then

Reply:
I suspect either original-developer caution, or something that was required before a refactor. Have moved it up to regular imports and things still work, so \o/